music_assistant/server/providers/plex/__init__.py

```diff
@@ -0,0 +1,467 @@
+from plexapi.audio import Album as PlexAlbum
+from plexapi.audio import Artist as PlexArtist
+from plexapi.audio import Playlist as PlexPlaylist
+from plexapi.audio import Track as PlexTrack
+from plexapi.library import MusicSection as PlexMusicSection
+from plexapi.media import AudioStream as PlexAudioStream
+from plexapi.media import Media as PlexMedia
+from plexapi.media import MediaPart as PlexMediaPart
+from plexapi.myplex import MyPlexAccount
+from plexapi.server import PlexServer
+
+from music_assistant.common.models.config_entries import ConfigEntry, ProviderConfig
+from music_assistant.common.models.enums import (
+    ConfigEntryType,
+    ContentType,
+    ImageType,
+    MediaType,
+    ProviderFeature,
+)
+from music_assistant.common.models.errors import InvalidDataError, LoginFailed, MediaNotFoundError
+from music_assistant.common.models.media_items import (
+    Album,
+    Artist,
+    MediaItemImage,
+    Playlist,
+    ProviderMapping,
+    SearchResults,
+    StreamDetails,
+    Track,
+)
+from music_assistant.common.models.provider import ProviderManifest
+from music_assistant.server import MusicAssistant
+from music_assistant.server.helpers.tags import parse_tags
+from music_assistant.server.models import ProviderInstanceType
+from music_assistant.server.models.music_provider import MusicProvider
+
+CONF_AUTH_TOKEN = "token"
+CONF_SERVER_NAME = "server"
+CONF_LIBRARY_NAME = "library"
+
+
+async def setup(
+    mass: MusicAssistant, manifest: ProviderManifest, config: ProviderConfig
+) -> ProviderInstanceType:
+    """Initialize provider(instance) with given configuration."""
+    prov = PlexProvider(mass, manifest, config)
+    await prov.handle_setup()
+    return prov
+
+
+async def get_config_entries(
+    mass: MusicAssistant, manifest: ProviderManifest  # noqa: ARG001
+) -> tuple[ConfigEntry, ...]:
+    """Return Config entries to setup this provider."""
+    return (
+        ConfigEntry(
+            key=CONF_SERVER_NAME, type=ConfigEntryType.STRING, label="Server", required=True
+        ),
+        ConfigEntry(
+            key=CONF_LIBRARY_NAME, type=ConfigEntryType.STRING, label="Library", required=True
+        ),
+        ConfigEntry(
+            key=CONF_AUTH_TOKEN, type=ConfigEntryType.SECURE_STRING, label="Token", required=True
+        ),
+    )
+
+
+class PlexProvider(MusicProvider):
+    """Provider for a plex music library."""
+
+    _plex_server: PlexServer = None
+    _plex_library: PlexMusicSection = None
+
+    async def handle_setup(self) -> None:
+        """Set up the music provider by connecting to the server."""
+        if not self.config.get_value(CONF_AUTH_TOKEN):
+            raise LoginFailed("Invalid login credentials")
+
+        def connect():
+            plex_account = MyPlexAccount(token=self.config.get_value(CONF_AUTH_TOKEN))
+            return plex_account.resource(self.config.get_value(CONF_SERVER_NAME)).connect()
+
+        self._plex_server = await self._run_async(connect)
```

Comment:
Is there some un-initialization needed when closing reloading the provider ?
In that case also implement a unload (async) method to do any cleanup

Reply:
I did not find any recommendations regarding un-initialization. As we don't provide a requests.Session object, the library creates its own session and should implicitly delete it